ksql-api/src/main/java/io/confluent/ksql/api/server/BlockingQueryPublisher.java

```diff
@@ -0,0 +1,175 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.server;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.api.spi.QueryPublisher;
+import io.confluent.ksql.query.RowConsumer;
+import io.vertx.core.Context;
+import io.vertx.core.WorkerExecutor;
+import java.util.Objects;
+import java.util.OptionalInt;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A query publisher that uses an internal blocking queue to store rows for delivery. It's currently
+ * necessary to use a blocking queue as Kafka Streams delivers message in a synchronous fashion with
+ * no back pressure. If the queue was not blocking then if the subscriber was slow the messages
+ * could build up on the queue eventually resulting in out of memory. The only mechanism we have to
+ * slow streams down is to block the thread. Kafka Streams uses dedicated streams per topology so
+ * this won't prevent the thread from doing useful work elsewhere but it does mean we can't have too
+ * many push queries in the server at any one time as we can end up with a lot of threads. Ideally
+ * Kafka Streams would use a non-blocking reactive model with back-pressure, e.g. using reactive
+ * streams. That way a small number of threads would be required to service all topologies and we
+ * wouldn't need to block.
+ */
+public class BlockingQueryPublisher extends BasePublisher<GenericRow>
+    implements QueryPublisher, RowConsumer {
+
+  private static final Logger log = LoggerFactory.getLogger(BlockingQueryPublisher.class);
+
+  public static final int SEND_MAX_BATCH_SIZE = 10;
+  public static final int BLOCKING_QUEUE_CAPACITY = 1000;
+
+  private final BlockingQueue<GenericRow> queue = new LinkedBlockingQueue<>(
+      BLOCKING_QUEUE_CAPACITY);
+  private final WorkerExecutor workerExecutor;
+  private PushQueryHandler queryHandle;
+  private String[] columnNames;
+  private String[] columnTypes;
+  private OptionalInt limit;
+  private int numAccepted;
+  private boolean complete;
+  private volatile boolean closed;
+
+  public BlockingQueryPublisher(final Context ctx,
+      final WorkerExecutor workerExecutor) {
+    super(ctx);
+    this.workerExecutor = Objects.requireNonNull(workerExecutor);
+  }
+
+  public void setQueryHandle(final PushQueryHandler queryHandle) {
+    this.queryHandle = Objects.requireNonNull(queryHandle);
+    this.limit = queryHandle.getLimit();
+    this.columnNames = queryHandle.getColumnNames();
+    this.columnTypes = queryHandle.getColumnTypes();
+  }
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP")
```

Comment:
+1

Don't expose mutable object state. It breaks encapsulation. Encapsulation is a pretty standard OO thing.

Reply:
I don't believe in dogmatic following of rules, sorry.

Reply:
This discussion is probably irrelevant now (https://github.com/confluentinc/ksql/pull/4495#discussion_r377852067) since you'll expose a `List` but here it goes anyway:

Rules are rules for a reason - I think unless there's a good reason not to follow them we probably should be. Specifically for this one, it's so easy to accidentally change the contents of an array, so why not prefer it to be an immutable view of the underlying data? This class is what "owns" the data (the column names/types) and it really doesn't expect them to change. Codifying it seem strictly non-negative (we definitely don't lose anything making it immutable, and we possibly lose by making it mutable).